### DISCUSSION TITLE
Issues connecting a Pixhawk 4 GPS module directly to a Raspberry Pi

### DISCUSSION THREAD

**Discussion by montmejat** [semuconsulting/pyubx2#135] [Q&A] [answered]
I'm trying to use a [Pixhawk 4 GPS module](https://fccid.io/2AP22-HG-C05/User-Manual/User-Manual-3889138) with a NEO-M8N directly from a Raspberry Pi (or Jetson).

I have connected the 5V pin, RX to TX, TX to RX and ground. The GPS blinks every second showing signs of life and I have activated the serial communication on my Raspberry.

I can kind of read stuff:

```
$ picocom /dev/ttyS0
picocom v3.1

port is        : /dev/ttyS0
flowcontrol    : none
baudrate is    : 9600
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        :
omap is        :
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready
'��F���x;����y�A�H��ӓ�
                      ��@���x
                             � ��
                                 ������0
                                        � �������
                                                 ��Ó�ł}=����ۂy��ӓ�~�������
                                                                          �S�� ��
                                                                                 �
                                                                                  ��C�����
                                                                                          ��e�����y�����������قY���7���ۂY
```

But I'm unable to parse the data:

```python
from serial import Serial
from pyubx2 import UBXReader

stream = Serial('/dev/ttyS0', 9600, timeout=3)
ubr = UBXReader(stream)
(raw_data, parsed_data) = ubr.read()
print(parsed_data)
```

Always returns `<UNKNOWN PROTOCOL(header=b'$!')>`.

Same thing happens on my Jetson Xavier NX. I tried using a baud rate of 38400. Any ideas on this? :)

**Comment by semuadmin** [comment, marked as answer]
Hi @aurelien-m  

I'm assuming you're using something like [this GPS module](https://docs.px4.io/v1.9.0/en/gps_compass/), but I'm unclear which serial port you're actually monitoring - the Pixhawk's or the RPi's?

What do you see if you simply type `screen /dev/ttyS0` (_or `cat /dev/ttyS0`_) at the terminal, without any intervening software?

Firstly, have you configured your Pixhawk module to pass-through NMEA or UBX on the designated serial port? To the best of my knowledge it doesn't do this by default. You'll need to refer to the vendor's or manufacturer's specifications for configuration details. I'm not in a position to provide support for your specific device.

Secondly, check that the pass-through baud rate is correct. The default UART baud rate on u-blox devices is 38400, but the Pixhawk may be configured differently - again, check the specifications.


Once you've confirmed that you are actually getting a valid NMEA or UBX stream on the serial port, try this updated code snippet:

```python
from serial import Serial
from pyubx2 import UBXReader

stream = Serial('/dev/ttyS0', 9600, timeout=3)
ubr = UBXReader(stream)
for raw_data, parsed_data in ubr:
   print(parsed_data)
```

**Comment by montmejat** [comment]
It works now! I had to use the ublox center software to reset its settings and now I'm able to read the data. I'm not sure what was wrong with it but it seems to work fine now. Thanks for your time!

**Reply by semuadmin** [reply]
OK glad you got it sorted.

FYI if your RPI is running a GUI, you could deploy [PyGPSClient](https://github.com/semuconsulting/PyGPSClient) on it to configure your M8N device, as an alternative to the Windows-only u-center software.

**Reply by montmejat** [reply]
Okay, I usually use Linux so that might be a nice alternative for me. Thanks!
